### PR TITLE
#12 add support to authentification

### LINKED ISSUE #12: add support to authentification
I will suport to authentication for RsConnectionPool + tests

### PR DESCRIPTION
**Pull request by CafeKrem**
fix #12

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -38,3 +38,14 @@ Metacello new
   repository: 'github://mumez/RediStick/repository';
   load: #('Core' 'ConnectionPool').
 ```
+
+## how to setUp test environement
+
+in this repository there is a redis.conf file , redis use it for initialisation
+
+```bash
```
Comment (CafeKrem) [inline README.md:46]: I decide to not setup automatically a test environment , I suppose that the user will set up by himself the Redis database

File: redis.conf [hunk 2]
```diff
@@ -0,0 +1,2 @@
+requirepass goodPassword
```
Comment (CafeKrem) [inline redis.conf:1]: the user will use this Redis configuration file to set up the Redis database for test

File: repository/RediStick-Connection-Pool.package/RsRedisConnectionPool.class/instance/userName.password..st [hunk 3]
```diff
@@ -0,0 +1,4 @@
+accessing
+userName: anUndefinedObject password: aString
```
Comment (CafeKrem) [inline repository/RediStick-Connection-Pool.package/RsRedisConnectionPool.class/instance/userName.password..st:2]: I will rename this argument into 'userName' , sorry my bad

Comment (CafeKrem) [reply repository/RediStick-Connection-Pool.package/RsRedisConnectionPool.class/instance/userName.password..st]: fix with commit 3793d52

File: repository/RediStick-Connection-Pool.package/RsRedisConnectionPool.class/instance/createStickAt..st [hunk 4]
```diff
@@ -6,4 +6,5 @@ createStickAt: key
 	stick := RsRediStick targetUrl: self primaryUrl.
 	stick beSticky.
 	stick connect.
+	self authentificateStick: stick.
```
Comment (mumez) [inline repository/RediStick-Connection-Pool.package/RsRedisConnectionPool.class/instance/createStickAt..st:9]: I think `stick authenticate` would be better.
Stick itself should have authentication feature. Connection pool's responsibility is keeping connections.

File: repository/RediStick-Core-Tests.package/RsRedisEndpointTest.class/instance/tearDown.st [hunk 5]
```diff
@@ -1,4 +1,5 @@
 running
 tearDown
+	self sendGoodPassword.
```
Comment (mumez) [inline repository/RediStick-Core-Tests.package/RsRedisEndpointTest.class/instance/tearDown.st:3]: Is this *always* needed for testing basic features of redis client? 
I think authentication would be an option. 
So, there should be RsRedisClientTestCase >> testAuthenticate and RsRedisClientTestCase >> testAuthenticateFailed.

If you combine them to setup/tearDown, it is difficult to say what features are tested.

File: repository/RediStick-Core.package/RsRedisEndpoint.class/instance/hMSet.dictionary..st [hunk 6]
```diff
@@ -2,5 +2,5 @@ commands-hashes
 hMSet: aKey dictionary: aDictionary
 	^ self
 		unifiedCommand:
-			{'HMSET'.
-			aKey} , (self keysAndValuesInlined: aDictionary)
\ No newline at end of file
+			{'HMSET'. aKey} , (self keysAndValuesInlined: aDictionary)
```
Comment (mumez) [inline repository/RediStick-Core.package/RsRedisEndpoint.class/instance/hMSet.dictionary..st:5]: It would be better to submit this fix as another PR. It is not related to authentication.

File: repository/RediStick-Core-Tests.package/RsRedisEndpointTest.class/instance/tearDown.st (hunk 5 above)
Comment (CafeKrem) [reply repository/RediStick-Core-Tests.package/RsRedisEndpointTest.class/instance/tearDown.st:3]: thank for this review, I misunderstand something when I wrote this.

Comment (CafeKrem) [reply repository/RediStick-Core-Tests.package/RsRedisEndpointTest.class/instance/tearDown.st]: I fix it in commit 
4144074

Comment (CafeKrem) [reply repository/RediStick-Core-Tests.package/RsRedisEndpointTest.class/instance/tearDown.st]: "I think authentication would be an option."

I must authenticate the stick because I suppose that the user (dev ) will deploy a Redis server using my redis.conf file (adding a password). if I not authenticate all tests will turn to red.

File: repository/RediStick-Connection-Pool.package/RsRedisConnectionPool.class/instance/createStickAt..st (hunk 4 above)
Comment (CafeKrem) [reply repository/RediStick-Connection-Pool.package/RsRedisConnectionPool.class/instance/createStickAt..st]: commit 940d78a

File: repository/RediStick-Core.package/RsRediStick.class/instance/connect.st [hunk 7]
```diff
@@ -0,0 +1,4 @@
+actions
+connect
```
Comment (CafeKrem) [inline repository/RediStick-Core.package/RsRediStick.class/instance/connect.st:2]: in my opinion, I think when we connect a stick it authenticate himself

Comment (mumez) [reply repository/RediStick-Core.package/RsRediStick.class/instance/connect.st:2]: The API would be simpler. However, there is a case that you do not need to authenticate.
So I think it would be better to call #shoudAuthenticate testing method explicitly.
```
connect
  super connect.
  self shoudAuthenticate ifTrue: [self authenticate].

shoudAuthenticate
  ^ self password notNil

authenticate
  self userName ifNil: [ ^ self endpoint auth: self password ]
  self endpoint auth: self password user: self userName ]
```

And 'authentificate' is a typo. 'Authentification' is valid as a noun, but as a verb, 'authentificate' is not common.

Comment (CafeKrem) [reply repository/RediStick-Core.package/RsRediStick.class/instance/connect.st:2]: thank you for this great reviewing, I learn a lot! 
you're right it's looking more sophisticated than my version. 
I will fix this in a next commit

Comment (CafeKrem) [reply repository/RediStick-Core.package/RsRediStick.class/instance/connect.st:2]: fix with commit 
27c5443

File: README.md (hunk 1 above)
Comment (CafeKrem) [reply README.md:46]: I think we can propose those 2 solutions (with/without docker ) , I will update readme in a next commit

Comment (CafeKrem) [reply README.md:70]: fix with commit  36133d3 and  2bdbdd4

**Comment by CafeKrem** [comment]
"The PR still includes the other changes that are unrelated to authentication like HSET testing"
commit 713c113

**Comment by CafeKrem** [comment]
for metadata file I think it's iceberg who deletes it. 

proof: 
https://github.com/CafeKrem/RediStick/commit/2bee65193aae9ee7af3f9678749c0de4adc0acf0
I try with a new branch from your dev branch, and you can see that some metadata file is removed

File: repository/RediStick-Connection-Pool-Tests.package/RsRedisConnectionPoolTest.class/instance/initializeConnectionPool.st [hunk 8]
```diff
@@ -0,0 +1,4 @@
+running
+initializeConnectionPool
+	RsRedisConnectionPool initialize.
+	connectionPool := RsRedisConnectionPool default
```
Comment (mumez) [inline repository/RediStick-Connection-Pool-Tests.package/RsRedisConnectionPoolTest.class/instance/initializeConnectionPool.st:4]: In this case, `RsRedisConnectionPool new` would be better. By using this, we can keep the actual (default) connection pool intact even when the test case stops with some unexpected error.

File: repository/RediStick-Connection-Pool.package/RsRedisConnectionPool.class/instance/createStickAt..st [hunk 9]
```diff
@@ -5,5 +5,6 @@ createStickAt: key
 	| stick |
 	stick := RsRediStick targetUrl: self primaryUrl.
 	stick beSticky.
+	stick username: userName password: password.
```
Comment (mumez) [inline repository/RediStick-Connection-Pool.package/RsRedisConnectionPool.class/instance/createStickAt..st:8]: Avoid direct access to the variables.
`stick username: self userName password: self password`

Comment (CafeKrem) [reply repository/RediStick-Connection-Pool.package/RsRedisConnectionPool.class/instance/createStickAt..st]: done a042c73   & f0cd103

File: repository/RediStick-Connection-Pool-Tests.package/RsRedisConnectionPoolTest.class/instance/initializeConnectionPool.st (hunk 8 above)
Comment (CafeKrem) [reply repository/RediStick-Connection-Pool-Tests.package/RsRedisConnectionPoolTest.class/instance/initializeConnectionPool.st]: done 66f4d7d

**Comment by CafeKrem** [comment]
is ok for you?

File: repository/RediStick-Core.package/RsRediStick.class/instance/shoudAuthenticate.st [hunk 10]
```diff
@@ -0,0 +1,3 @@
+actions
+shoudAuthenticate
+	^ password isNotNil
```
Comment (mumez) [inline repository/RediStick-Core.package/RsRediStick.class/instance/shoudAuthenticate.st:3]: `password notNil` is better. `isNotNil` is Pharo specific and loses portability to other smalltalk dialects.

File: repository/RediStick-Core.package/RsRedisEndpoint.class/instance/mSet..st [hunk 11]
```diff
@@ -1,3 +1,4 @@
 commands-strings
 mSet: aDictionary
-	^ self unifiedCommand: {'MSET'} , (self keysAndValuesInlined: aDictionary)
\ No newline at end of file
+	^ self unifiedCommand: {'MSET'} , (self keysAndValuesInlined: aDictionary)
```
Comment (mumez) [inline repository/RediStick-Core.package/RsRedisEndpoint.class/instance/mSet..st:3]: MSET tests are also unrelated to the PR topic.

Comment (CafeKrem) [reply repository/RediStick-Core.package/RsRedisEndpoint.class/instance/mSet..st]: done 
7dc7ee4

**Comment by mumez** [comment]
You can manually salvage unintentionally deleted properties files from the current develop branch and push the PR again.  We are getting close to the merge!

**Comment by mumez** [comment]
In the future, I will change the file format from Cypress to Tonel, so it would better fit to Iceberg. However, RediStick is still used with some legacy (pharo 6) projects, so we need some time for migration.

**Comment by CafeKrem** [comment]
"You can manually salvage unintentionally deleted properties files from the current develop branch and push the PR again."
but how I can do this?

**Comment by CafeKrem** [comment]
I mean via git command

**Comment by CafeKrem** [comment]
done

**Comment by mumez** [comment]
> "You can manually salvage unintentionally deleted properties files from the current develop branch and push the PR again."
> but how I can do this?

You can cherry-pick via git command, but a simpler way is just to clone the develop branch again to other directory (like RediStick-develop) and copy some unintentionally deleted files to your working directory.

I think these files are:
```
RediStick-Connection-Pool.package/RsRedis.class/methodProperties.json 
RediStick-Connection-Pool.package/RsRedisConnectionPool.class/methodProperties.json 
RediStick-Connection-Pool.package/RsRedisProxy.class/methodProperties.json
```

Thank you for your patience. I would like to keep the PR clean.

File: repository/RediStick-Connection-Pool-Tests.package/RsRedisConnectionPoolTest.class/instance/setUp.st [hunk 12]
```diff
@@ -0,0 +1,4 @@
+running
+setUp
+	self initializeConnectionPool.
+	connectionPool userName: nil password: self goodPassword
```
Comment (mumez) [inline repository/RediStick-Connection-Pool-Tests.package/RsRedisConnectionPoolTest.class/instance/setUp.st:4]: How about adding `connectionPool password: self goodPassword` short-cut method?

File: repository/RediStick-Connection-Pool.package/RsRedisConnectionPool.class/instance/createStickAt..st [hunk 13]
```diff
@@ -5,5 +5,6 @@ createStickAt: key
 	| stick |
 	stick := RsRediStick targetUrl: self primaryUrl.
 	stick beSticky.
+	stick username: self userName password: self password.
```
Comment (mumez) [inline repository/RediStick-Connection-Pool.package/RsRedisConnectionPool.class/instance/createStickAt..st:8]: `username:` should be `userName:` for consistency.

File: repository/RediStick-Core.package/RsRediStick.class/instance/username.password..st [hunk 14]
```diff
@@ -0,0 +1,4 @@
+setter
+username: aUsername password: aPassword 
+	self userName: aUsername.
```
Comment (mumez) [inline repository/RediStick-Core.package/RsRediStick.class/instance/username.password..st:3]: In this case, I prefer direct variable access, because this method is just an initializer.
```
userName := aUserName.
password := aPassword.
````

File: repository/RediStick-Core.package/RsRedisEndpoint.class/instance/unifiedCommand.withDictionary..st [hunk 15]
```diff
@@ -0,0 +1,3 @@
+private
+unifiedCommand: anArray withDictionary: aDictionary
```
Comment (mumez) [inline repository/RediStick-Core.package/RsRedisEndpoint.class/instance/unifiedCommand.withDictionary..st:2]: I think this method had no caller, maybe this should be included in other PR.

**Review by mumez — COMMENTED** [review]
Almost done. Please give the finishing touch

File: repository/RediStick-Core.package/RsRedisEndpoint.class/instance/unifiedCommand.withDictionary..st (hunk 15 above)
Comment (CafeKrem) [reply repository/RediStick-Core.package/RsRedisEndpoint.class/instance/unifiedCommand.withDictionary..st]: fix 573893d

File: repository/RediStick-Core.package/RsRediStick.class/instance/username.password..st (hunk 14 above)
Comment (CafeKrem) [reply repository/RediStick-Core.package/RsRediStick.class/instance/username.password..st]: fix 35150ae

File: repository/RediStick-Connection-Pool.package/RsRedisConnectionPool.class/instance/createStickAt..st (hunk 13 above)
Comment (CafeKrem) [reply repository/RediStick-Connection-Pool.package/RsRedisConnectionPool.class/instance/createStickAt..st:8]: fix 35150ae

File: repository/RediStick-Connection-Pool-Tests.package/RsRedisConnectionPoolTest.class/instance/setUp.st (hunk 12 above)
Comment (CafeKrem) [reply repository/RediStick-Connection-Pool-Tests.package/RsRedisConnectionPoolTest.class/instance/setUp.st:4]: yes we can do it

Comment (CafeKrem) [reply repository/RediStick-Connection-Pool-Tests.package/RsRedisConnectionPoolTest.class/instance/setUp.st:4]: fix 23db754

**Comment by mumez** [comment]
Thanks for the updates. Unfortunately, many methodProperties files were deleted again. Please salvage them, so I can finally merge. 
https://github.com/mumez/RediStick/pull/14/files

**Comment by mumez** [comment]
OK. I will try to salvage deleted files by myself.  #develop branch will be unstable for a while but recovered soon.
